src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectMethodExecutor.cs

```diff
@@ -87,5 +103,151 @@ private static ActionExecutor WrapVoidAction(VoidActionExecutor executor)
+                }
+
+                var taskValueType = GetTaskInnerTypeOrNull(returnType);
+                if (taskValueType != null)
+                {
+                    // for: public Task<T> Action()
+                    // constructs: return (Task<object>)Convert<T>((Task<T>)result)
+                    var genericMethodInfo = _convertOfTMethod.MakeGenericMethod(taskValueType);
+                    var genericMethodCall = Expression.Call(null, genericMethodInfo, castMethodCall);
+                    var convertedResult = Expression.Convert(genericMethodCall, typeof(Task<object>));
+                    return convertedResult;
+                }
+
+                // This will be the case for types which have derived from Task and Task<T>
+                throw new InvalidOperationException(Resources.FormatActionExecutor_UnexpectedTaskInstance(
+                    methodInfo.Name,
+                    methodInfo.DeclaringType));
+            }
+
+            // Action methods which use dynamic keyword but return a Task or Task<T> will also be coerced to Task<object>
+            Expression<Func<object, Task<object>>> createTaskFromResultExpression = ((result) => Task.FromResult(result));
+            var createTaskFromResultMethodInfo = ((MethodCallExpression)createTaskFromResultExpression.Body).Method;
+            return Expression.Call(null, createTaskFromResultMethodInfo, castMethodCall);
+        }
+
+        private static ActionExecutorAsync WrapVoidActionAsync(VoidActionExecutor executor)
+        {
+            return delegate (object target, object[] parameters)
+            {
+                executor(target, parameters);
+                return Task.FromResult<object>(null);
+            };
+        }
+        private static void ThrowIfWrappedTaskInstance(Type actualTypeReturned, string methodName, Type declaringType)
+        {
+            // Throw if a method declares a return type of Task and returns an instance of Task<Task> or Task<Task<T>>
+            // This most likely indicates that the developer forgot to call Unwrap() somewhere.
+            if (actualTypeReturned != typeof(Task))
+            {
+                var innerTaskType = GetTaskInnerTypeOrNull(actualTypeReturned);
+                if (innerTaskType != null && typeof(Task).IsAssignableFrom(innerTaskType))
+                {
+                    throw new InvalidOperationException(
+                        Resources.FormatActionExecutor_WrappedTaskInstance(
+                            methodName,
+                            declaringType,
+                            actualTypeReturned.FullName));
+                }
+            }
+        }
+
+        private static Task<object> CoerceTaskType(object result, string methodName, Type declaringType)
+        {
+            var resultAsTask = result as Task;
+            ThrowIfWrappedTaskInstance(resultAsTask.GetType(), methodName, declaringType);
```

Comment:
If you're going to keep this around, let's remeasure. I'm concerned we might want to get rid of this too.